monix-tail/shared/src/main/scala/monix/tail/internal/IterantFromReactivePublisher.scala

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2014-2018 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.tail.internal
+
+import cats.effect.{Async, Timer}
+import cats.syntax.apply._
+import monix.execution.misc.AsyncVar
+import monix.execution.schedulers.TrampolineExecutionContext.immediate
+import monix.tail.{Iterant, PullStrategy}
+import monix.tail.Iterant.{Halt, Next, NextBatch}
+import monix.tail.PullStrategy.Batched
+import monix.tail.batches.Batch
+import org.reactivestreams.{Publisher, Subscriber, Subscription}
+
+private[tail] object IterantFromReactivePublisher {
+  class UnfoldSubscriber[F[_], A] (bufferSize: Int)(implicit
+    F: Async[F],
+    timer: Timer[F]
+  )
+    extends Subscriber[A]
+  {
+    private[this] val result = AsyncVar.empty[Iterant[F, A]]
+    private[this] var subscription: Subscription = _
+    private[this] var offset = 0
+    private[this] var buffer: Array[Any] =
+      if (bufferSize == 1) null else new Array(bufferSize)
```

Comment:
I'm using array here for memory efficiency, but it might not be the best idea if user passes in a buffer size of `Int.MaxValue`. Maybe `ArrayBuffer` is a better solution, at least if `bufferSize` is greater than some reasonably small number.